### PR TITLE
Add CircuitBreaker to TDigest, Step 1: Raw arrays to Arrays wrapper

### PR DESCRIPTION
**Pull request by ivancea**
Part of https://github.com/elastic/elasticsearch/issues/99815

## Steps
1. Migrate TDigest classes to use a custom Array implementation. Temporarily use a simple array wrapper (This PR)
2. Implement a BigArrays class and replace the wrapper with it. Add Releasable/AutoCloseable and ensure everything is closed
3. Account remaining TDigest classes size

Every step should be safely mergeable to main:
- The first one should have no impact.
- The second and third ones will start increasing the CB count partially.

## Considerations
The third step will probably require some other interface to manually count used memory _before_ creation of the classes. Something like a `TDigestCircuitBreaker`.
After building this one, I've started considering if it would make sense to just do the breaker, and not migrate things to BigArrays. Simply call a `breaker.increase(...)` before array creation.

The pros I see of BigArrays:
- Automatically verified in ESTestCases. Which means tests will fail unless everything is correctly `.close()`d
- Automatic byte counting, without added calculations to the TDigests (Apart of the close() ones)

And the cons:
- Filling the code with .get()/.set()
- `.sort()` will require a custom implementation for BigArrays. Same for the `.set()` method that copies a range of an array to another

## Benchmarks
This is a comparison of the benchmarks between main (left) and this branch (right). Bigger is worse:
```
TDigest
(compression)  (distribution)  (tdigestFactory)  Score   Error  ->  Score   Error  Units
          100          NORMAL             MERGE  0,157 ± 0,248  ->  0,170 ± 0,071  us/op
          100          NORMAL          AVL_TREE  0,263 ± 0,078  ->  0,309 ± 0,170  us/op
          100          NORMAL            HYBRID  0,167 ± 0,040  ->  0,169 ± 0,042  us/op
          100        GAUSSIAN             MERGE  0,159 ± 0,041  ->  0,163 ± 0,070  us/op
          100        GAUSSIAN          AVL_TREE  0,339 ± 0,127  ->  0,336 ± 0,029  us/op
          100        GAUSSIAN            HYBRID  0,163 ± 0,049  ->  0,167 ± 0,072  us/op
          300          NORMAL             MERGE  0,174 ± 0,044  ->  0,183 ± 0,031  us/op
          300          NORMAL          AVL_TREE  0,443 ± 0,084  ->  0,438 ± 0,079  us/op
          300          NORMAL            HYBRID  0,180 ± 0,059  ->  0,184 ± 0,039  us/op
          300        GAUSSIAN             MERGE  0,167 ± 0,040  ->  0,173 ± 0,054  us/op
          300        GAUSSIAN          AVL_TREE  0,403 ± 0,098  ->  0,387 ± 0,125  us/op
          300        GAUSSIAN            HYBRID  0,183 ± 0,031  ->  0,178 ± 0,049  us/op
```

```
StableSort
(sortDirection)   Score   Error  ->   Score   Error Units
              0  16,435 ± 1,443  ->  15,909 ± 0,745 ms/op
              1   5,237 ± 0,184  ->   4,994 ± 0,461 ms/op
             -1   5,458 ± 0,398  ->   4,696 ± 0,265 ms/op
```

There's barely any relevant affectation I can see.

### CODE REVIEW

File: libs/tdigest/src/test/java/org/elasticsearch/tdigest/SortTests.java
Comment (ivancea) [inline libs/tdigest/src/test/java/org/elasticsearch/tdigest/SortTests.java:1]: The deleted tests here were testing the Sort method with 3 array inputs: Key, value and weight. That method was removed as it wasn't being used, and so the dependent tests.

Other tests were updated to use stableSort() instead, which is the currently used method

File: libs/tdigest/src/test/java/org/elasticsearch/tdigest/TDigestTests.java [hunk 1]
```diff
@@ -542,4 +544,8 @@ public void testMonotonicity() {
             lastQuantile = q;
         }
     }
+
+    protected static TDigestArrays arrays() {
+        return WrapperTDigestArrays.INSTANCE;
```
Comment (ivancea) [inline libs/tdigest/src/test/java/org/elasticsearch/tdigest/TDigestTests.java:549]: In the next PR, I'll change this to the real BigArrays implementation

File: server/src/main/java/org/elasticsearch/common/util/BigArrays.java [hunk 2]
```diff
@@ -470,13 +471,13 @@ public T set(long index, T value) {
     private final BigArrays circuitBreakingInstance;
     private final String breakerName;
 
-    public BigArrays(PageCacheRecycler recycler, @Nullable final CircuitBreakerService breakerService, String breakerName) {
+    public BigArrays(@Nullable PageCacheRecycler recycler, @Nullable final CircuitBreakerService breakerService, String breakerName) {
```
Comment (ivancea) [inline server/src/main/java/org/elasticsearch/common/util/BigArrays.java:474]: Unrelated here, but this is nullable, so fixed it

File: benchmarks/src/main/java/org/elasticsearch/benchmark/tdigest/SortBench.java [hunk 3]
```diff
@@ -58,22 +60,22 @@ public class SortBench {
...
-    public void quicksort() {
-        int[] order = new int[size];
+    public void stableSort() {
+        TDigestIntArray order = WrapperTDigestArrays.INSTANCE.newIntArray(size);
         for (int i = 0; i < size; i++) {
-            order[i] = i;
+            order.set(i, i);
         }
-        Sort.sort(order, values, null, values.length);
+        Sort.stableSort(order, values, values.size());
```
Comment (nik9000) [inline benchmarks/src/main/java/org/elasticsearch/benchmark/tdigest/SortBench.java:79]: Did this spit out different performance numbers before and after?

File: libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/WrapperTDigestArrays.java [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+import org.apache.lucene.util.ArrayUtil;
+
+import java.util.Arrays;
+
+/**
+ * Temporal TDigestArrays with raw arrays.
+ *
+ * <p>
+ *     Delete after the right implementation for BigArrays is made.
+ * </p>
```
Comment (nik9000) [inline libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/WrapperTDigestArrays.java:33]: I think we'd move this implementation to the tests rather than delete.

File: libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/WrapperTDigestArrays.java [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+ */
+public class WrapperTDigestArrays implements TDigestArrays {
```
Comment (nik9000) [inline libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/WrapperTDigestArrays.java:35]: I'm not sure I'd call this `Wrapper` - it's more a `NativeJdkTDigestArrays` - but the name doesn't make a big difference if we're moving it to tests eventually.

File: libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/TDigestLongArray.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * This project is based on a modification of https://github.com/tdunning/t-digest which is licensed under the Apache 2.0 License.
+ */
+
+package org.elasticsearch.tdigest.arrays;
+
+/**
+ * Minimal interface for LongArray-like classes used within TDigest.
+ */
+public interface TDigestLongArray extends AutoCloseable {
```
Comment (nik9000) [inline libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/TDigestLongArray.java:27]: If you depend on elasticsearch-core you can have `Releasable` which just can't throw anything.

File: libs/tdigest/src/test/java/org/elasticsearch/tdigest/TDigestTests.java (hunk 1 above)
Comment (nik9000) [reply libs/tdigest/src/test/java/org/elasticsearch/tdigest/TDigestTests.java:549]: It'd be nice, at least in tests, to track the unclosed stuff. I think you could try using `LeakTracker`.

Comment (nik9000) [reply libs/tdigest/src/test/java/org/elasticsearch/tdigest/TDigestTests.java:549]: But that can wait for another PR if you want.

**Review by nik9000 — COMMENTED** [review]
That's how I'd do it, yeah. It's possible to use this approach in a way that there's no performance overhead on the jdk - at least no overhead on read and write - there's some overhead on coordinating the shared resources and stuff. But that's good to me.

File: libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/WrapperTDigestArrays.java (hunk 4 above)
Comment (ivancea) [reply libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/WrapperTDigestArrays.java:33]: Was thinking of that, but the BigArrays iml with MockBigArrays has the side effect of checking for leaks. So it feels better to me to just keep that implementation. But I'll consider it in the next PR

File: benchmarks/src/main/java/org/elasticsearch/benchmark/tdigest/SortBench.java (hunk 3 above)
Comment (nik9000) [reply benchmarks/src/main/java/org/elasticsearch/benchmark/tdigest/SortBench.java:79]: If you run the thing before and after you can make a little chart like I did in https://github.com/elastic/elasticsearch/pull/112717. It feels nice to have numbers to look at stored in github.

File: libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/WrapperTDigestArrays.java (hunk 4 above)
Comment (nik9000) [reply libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/WrapperTDigestArrays.java:33]: Ah, yeah, we could indeed just do that.

File: libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/TDigestLongArray.java (hunk 6 above)
Comment (nik9000) [reply libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/TDigestLongArray.java]: :+1:

File: benchmarks/src/main/java/org/elasticsearch/benchmark/tdigest/SortBench.java (hunk 3 above)
Comment (ivancea) [reply benchmarks/src/main/java/org/elasticsearch/benchmark/tdigest/SortBench.java:79]: Nice, gonna do that. I guess you did the "->" manually right? In case there's a simpler way to do that

Comment (ivancea) [reply benchmarks/src/main/java/org/elasticsearch/benchmark/tdigest/SortBench.java:79]: Added the 2 relevant benchmarks (TDigest and Sort) to the description

**Comment by ivancea** [comment]
@elasticmachine update branch

File: benchmarks/src/main/java/org/elasticsearch/benchmark/tdigest/SortBench.java (hunk 3 above)
Comment (nik9000) [reply benchmarks/src/main/java/org/elasticsearch/benchmark/tdigest/SortBench.java:79]: > Nice, gonna do that. I guess you did the "->" manually right? In case there's a simpler way to do that

Yes. I stuff both examples into vscode and grab the output columns from one and stick them into another. I tend to trim a few other bits of the output while I'm there.

**Comment by nik9000** [comment]
> There's barely any relevant affectation I can see.

That indeed looks like noise. :+1:

File: libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/TDigestDoubleArray.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Resizes the array. If the new size is bigger than the current size, the new elements are set to 0.
+     */
+    void resize(int newSize);
+
+    /**
+     * Copies {@code len} elements from {@code buf} to this array.
+     * <p>
+     *     Copy must be made in reverse order. That is, starting from offset+len-1 to offset.
+     *     This is, because it will be used to copy an array to itself.
```
Comment (kkrik-es) [inline libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/TDigestDoubleArray.java:47]: Nit: this statement is somewhat misleading, unless I'm mistaken. Are you referring to a specific case where self-copies are safe? If so, let's clearly document it and add asserts for it.

File: libs/tdigest/src/test/java/org/elasticsearch/tdigest/SortTests.java [hunk 8]
```diff
@@ -92,58 +92,10 @@ public void testRepeated() {
             values[i] = Math.rint(10 * ((double) i / n)) / 10.0;
         }
 
-        Sort.sort(order, values, null, values.length);
+        sort(order, values, values.length);
         checkOrder(order, values);
     }
 
-    public void testRepeatedSortByWeight() {
```
Comment (kkrik-es) [inline libs/tdigest/src/test/java/org/elasticsearch/tdigest/SortTests.java:99]: Why is this removed?

Comment (kkrik-es) [reply libs/tdigest/src/test/java/org/elasticsearch/tdigest/SortTests.java:99]: Ah just tests sorting, sounds good.

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/TDigestState.java [hunk 9]
```diff
@@ -47,23 +49,35 @@ static Type valueForHighAccuracy() {
 
     private final Type type;
 
+    // TODO: DELETE ME
```
Comment (kkrik-es) [inline server/src/main/java/org/elasticsearch/search/aggregations/metrics/TDigestState.java:52]: Is this still used?

Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/TDigestState.java:52]: I saw the comment.. Nit: let's be specific about when this can be deleted (e.g. once there are no more references in tests?)

**Review by kkrik-es — APPROVED** [review]
Iiuc there's no perf penalty from the interface wrapping array ops, so far so good.

File: libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/TDigestDoubleArray.java (hunk 7 above)
Comment (ivancea) [reply libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/TDigestDoubleArray.java]: I extracted those interfaces from existing TDigest usecases. Very specific, avoiding dead code.

This comes from a `System.arraycopy(src, srcIndex, dest, destIndex, len)`, which lets you have the same source and destination, and will ensure that it will be copied correctly in any case.
But as I didn't want to add dead code, I added that comment.

As to make it clearer, just pushed some improved comments (I believe), and some asserts so it's always called as it's expected to be called.
- For the Double array, just removed the comment and asserted that it can't be copied to itself
- For the Int array, asserted that if it's copied to itself, the source index should be greater

_Yeah, oddly specific..._

**Review by nik9000 — APPROVED** [review]
